src/api/app/controllers/webui/users/tokens_controller.rb

```diff
@@ -0,0 +1,84 @@
+class Webui::Users::TokensController < Webui::WebuiController
+  before_action :set_token, only: [:destroy]
+  before_action :set_params, :validate_package, :set_package, :set_scm_token, only: [:create]
+
+  after_action :verify_authorized, except: :index
+  after_action :verify_policy_scoped, only: :index
+
+  def index
+    @tokens = policy_scope([:webui, Token]).page(params[:page])
+  end
+
+  def new
+    @token = Token.new
+    authorize [:webui, @token]
+  end
+
+  def create
+    @token = Token.token_type(@params[:operation]).new(@params.merge(user: User.session!, package: @package, scm_token: @scm_token))
+
+    authorize [:webui, @token]
+
+    respond_to do |format|
+      format.js do
+        if @token.valid? && @token.save
+          render partial: 'create', locals: {
+            string: @token.string,
+            flash: { success: "Token successfully created! Make sure you save it - you won't be able to access it again." }
+          }
+        else
+          render partial: 'create', locals: {
+            flash: { error: "Failed to create token: #{@token.errors.full_messages.to_sentence}." }
+          }
+        end
+      end
+    end
+  end
+
+  def destroy
+    authorize [:webui, @token]
+    @token.destroy
+    flash[:success] = 'Token was successfully deleted.'
+    redirect_to tokens_url
+  end
+
+  private
+
+  def set_token
+    @token = Token.find(params[:id])
+  end
+
+  def set_params
+    @params = params.require(:token).except(:string_readonly).permit(:operation, :scm_token).tap do |token_parameters|
+      token_parameters.require(:operation)
+    end
+    @package_params = params.slice(:project_name, :package_name).permit!
+  end
+
+  def validate_package
+    return unless @package_params[:project_name].present? ^ @package_params[:package_name].present?
+
+    render partial: 'create', locals: {
+      flash: { error: 'Failed to create token: when providing an optional package, both project name and package name must be provided.' }
+    }
+  end
+
+  def set_package
+    return unless @package_params[:project_name].present? && @package_params[:package_name].present?
+
+    begin
+      @package = Package.get_by_project_and_name(@package_params[:project_name], @package_params[:package_name])
+    rescue Package::UnknownObjectError
+      render partial: 'create', locals: {
+        flash: { error: 'Failed to create token: when providing an optional package, the package must exist. '\
+                        "Package '#{@package_params[:project_name]}/#{@package_params[:package_name]}' does not exist." }
+      }
+    end
+  end
+
+  def set_scm_token
```

Comment:
This could go in the `initialize` method of `Token::Workflow` since it's only for that model.